File: exercism/csharp/pig-latin/PigLatin.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class PigLatin
{
    static string[] ConstStems = new string[] { "ch", "qu", "squ", "thr", "th", "sch" };
    static string[] VowelStems = new string[] { "yt", "xr", "a", "e", "i", "o", "u" };

    public static string Translate (string given)
    {
        return String.Join(" ", given.Split().Select(TranslateWord));
    }

    static string TranslateWord (string word)
    {
        string head, tail;
        foreach (var stem in ConstStems)
        {
            head = word.Substring(0, stem.Length);
            tail = word.Substring(stem.Length);
            if (head == stem) return tail + head + "ay";
        }
        foreach (var stem in VowelStems)
        {
            head = word.Substring(0, stem.Length);
            tail = word.Substring(stem.Length);
            if (head == stem) return head + tail + "ay";
        }
        return word.Substring(1) + word[0] + "ay";
    }
}

File: exercism/go/pig-latin/pig_latin.go
package igpay

import "strings"

// PigLatin converts the given string according to the pig latin rules
func PigLatin(phrase string) string {
	english := strings.Split(phrase, " ")
	latin := make([]string, len(english))
	for i, word := range english {
		isVowel := false
		n := 1
		for _, r := range rules {
			if head := word[:len(r.stem)]; head == r.stem {
				isVowel = r.isVowel
				n = len(r.stem)
				break
			}
		}
		if isVowel {
			latin[i] = word + "ay"
		} else {
			latin[i] = word[n:] + word[:n] + "ay"
		}
	}
	return strings.Join(latin, " ")
}

var rules = []struct {
	stem    string
	isVowel bool
}{
	{"ch", false}, {"qu", false}, {"squ", false}, {"thr", false}, {"th", false}, {"sch", false},
	{"yt", true}, {"xr", true}, {"a", true}, {"e", true}, {"i", true}, {"o", true}, {"u", true},
}
